src/backend/utils/adt/network.c

```diff
@@ -389,6 +412,413 @@ network_cmp(PG_FUNCTION_ARGS)
+	}
+
+	/*
+	 * Create an integer representation of the IP address by taking its first
+	 * 4 or 8 bytes. We take 8 bytes of an IPv6 address on a 64-bit machine
+	 * and 4 bytes on a 32-bit. Always take all 4 bytes of an IPv4 address.
+	 *
+	 * We're consuming an array of char, so make sure to byteswap on little
+	 * endian systems (an inet's IP array emulates big endian in that the
+	 * first byte is always the most significant).
+	 */
+	if (ip_family(authoritative) == PGSQL_AF_INET6)
+	{
+		ipaddr_datum = *((Datum *) ip_addr(authoritative));
+		ipaddr_datum = DatumBigEndianToNative(ipaddr_datum);
+	}
+	else
+	{
+		uint32		ipaddr_datum32 = *((uint32 *) ip_addr(authoritative));
+#ifndef WORDS_BIGENDIAN
+		ipaddr_datum = pg_bswap32(ipaddr_datum32);
+#endif
+	}
+
+	/*
+	 * Number of bits in subnet. e.g. An IPv4 that's /24 is 32 - 24 = 8.
+	 *
+	 * However, only some of the bits may have made it into the fixed sized
+	 * datum, so take the smallest number between bits in the subnet and bits
+	 * in the datum which are not part of the netmask.
+	 */
+	datum_subnet_size = Min(ip_maxbits(authoritative) - ip_bits(authoritative),
+							SIZEOF_DATUM * BITS_PER_BYTE - ip_bits(authoritative));
+
+	/* we may have ended up with < 0 for a large netmask */
+	if (datum_subnet_size <= 0)
+	{
+		/* the netmask occupies the entirety of datum */
+		subnet = (Datum) 0;
+		netmask = ipaddr_datum;
+	}
+	else
+	{
+		Assert(datum_subnet_size <= SIZEOF_DATUM * BITS_PER_BYTE);
+
+		/*
+		 * This shift creates a power of two like `0010 0000`, and subtracts
+		 * one to create a bitmask for an IP's subnet bits like `0001 1111`.
+		 *
+		 * Note that `datum_subnet_mask` may be == 0, in which case we'll
+		 * generate a 0 bitmask and `subnet` will also come out as 0.
+		 */
+		subnet_bitmask = (((Datum) 1) << datum_subnet_size) - 1;
+
+		/* bitwise AND the IP and bitmask to extract just the subnet bits */
+		subnet = ipaddr_datum & subnet_bitmask;
+
+		/* and likewise, use the mask's complement to get the netmask bits */
+		netmask = ipaddr_datum & ~subnet_bitmask;
+	}
+
+#if SIZEOF_DATUM == 8
+
+	if (ip_family(authoritative) == PGSQL_AF_INET6)
+	{
+		/*
+		 * IPv6 on a 64-bit machine: keep the most significant 63 netmasked
+		 * bits.
+		 */
+		res |= netmask >> 1;
+	}
+	else
+	{
+		/*
+		 * IPv4 on a 64-bit machine: keep all 32 netmasked bits, netmask size,
+		 * and most significant 25 subnet bits (see diagram above for more
+		 * detail).
+		 */
+
+		Datum		netmask_shifted;
+		Datum		netmask_size_and_subnet = 0;
+
+		/*
+		 * an IPv4 netmask has a maximum value of 32 which takes 6 bits to
+		 * contain
+		 */
+		Datum		netmask_size = (Datum) ip_bits(authoritative);
+
+		Assert(netmask_size <= ip_maxbits(authoritative));
+		netmask_size_and_subnet |= netmask_size << ABBREV_BITS_INET4_SUBNET;
+
+		/*
+		 * if we have more than 25 subnet bits of information, shift it down
+		 * to the available size
+		 */
+		if (datum_subnet_size > ABBREV_BITS_INET4_SUBNET)
```

Comment:
Why not move this up to where subnet is initially formed?

Reply:
Why not always shift when you do that, so the general rule is that subnet's most significant bit is always the actual most significant bit? Yeah, you only benefit from the shift when doing so makes space for less significant bits at the end, but I think it's clearer that way.

Reply:
Why not just start off with a subnet integer whose most significant bit is the actual most significant bit in all cases? That's simpler and easier to debug. It immediately becomes the component that you saw from the diagram.

Also, shifting is slightly cheaper than the if statement, because you eliminate a branch. Shifting isn't expensive at all.

Reply:
I think you're suggesting something like this?

``` c
subnet <<= ip_bits(authoritative);

// ... later

subnet >>= ((SIZE_OF_DATUM - BITS_PER_BYTE) - ABBREV_BITS_INET4_SUBNET);
```

Hm, I'll grant you that it's definitely an alternative, but personally, I don't think it makes the code easier to understand, and saving one if statement probably isn't worth it. I'm going to leave it for now.

Thanks for the additional review! Good stuff. I feel like this code is starting to look pretty decent.